music/src/ddsp/ddsp.ts

```diff
@@ -0,0 +1,387 @@
+
+      const ldClipped = ldAdjusted.clipByValue(LOWEST_LD, SELECTED_LD_AVG_MAX);
+
+      // rescale
+      const oldMin = ldClipped.min();
+      const subOldMin = ldMean.sub(oldMin);
+      return ldClipped
+        .sub(oldMin)
+        .div(subOldMin)
+        .mul(SELECTED_LD_MEAN - LOWEST_LD)
+        .add(LOWEST_LD);
+    });
+
+    const reshapedConf = tf.reshape(af.confidences, [-1, 1, 1]);
+    const smoothConf = tf.pool(
+      reshapedConf as Tensor3D,
+      [CONF_SMOOTH_SIZE, 1],
+      'avg',
+      'same'
+    );
+    const smoothConfReshaped = smoothConf.reshape([-1]);
+    const confMask = tf.lessEqual(smoothConfReshaped, CONF_THRESHOLD);
+
+    maskedPower = tf.tidy(() => {
+      const confReduction = confMask.mul(LD_CONF_REDUCTION);
+      const loudnessReduced = ldFinal.add(confReduction);
+      const finalMaskedPower = tf.maximum(loudnessReduced, LOWEST_LD);
+      confReduction.dispose();
+      loudnessReduced.dispose();
+      return finalMaskedPower;
+    });
+
+    loudness_db = (await maskedPower.array()) as number[];
+
+    ldShiftedGreater.dispose();
+    ldShifted.dispose();
+    ldFinal.dispose();
+    ldGreater.dispose();
+    smoothConf.dispose();
+    smoothConfReshaped.dispose();
+    reshapedConf.dispose();
+    confMask.dispose();
+  }
+  // tslint:disable-next-line: variable-name
+  let f0_hz: number[] = [];
+  if (af.f0_hz.length > 0) {
+    // shift the pitch register.
+    const _p = await hzToMidi(af.f0_hz);
+
+    const p = tf.tidy(() => {
+      for (let i = 0, n = _p.length; i < n; ++i) {
+        if (_p[i] === -Infinity) {
+          _p[i] = 0;
+        }
+      }
+
+      return _p;
+    });
+
+    const confMask = tf.lessEqual(af.confidences, CONF_THRESHOLD);
+    const maskedPowerThreshold = maskedPower.greater(SELECTED_LD_THRESH);
+    const mask = tf.logicalOr(confMask, maskedPowerThreshold);
+    const pMask = await tf.booleanMaskAsync(p, mask);
+    const pMean = pMask.mean();
+    const pDiff = await tf.sub(SELECTED_P_MEAN, pMean);
+    const pDiffVal = await pDiff.array();
+
+    const finalf0 = tf.tidy(() => {
+      let pDiffOctave = (pDiffVal as number) / 12;
+      pDiffOctave = Math.round(pDiffOctave);
+      const shiftedF0WithOctave = shiftF0(af.f0_hz, pDiffOctave);
+      return shiftedF0WithOctave;
+    });
+
+    f0_hz = (await finalf0.array()) as number[];
+
+    maskedPower.dispose();
+    pMean.dispose();
+    pMask.dispose();
+    pDiff.dispose();
+    finalf0.dispose();
+    maskedPowerThreshold.dispose();
+    mask.dispose();
+    confMask.dispose();
+  }
+
+  return { f0_hz, loudness_db };
+}
+
+export function getModelValues(checkpointUrl: string): ModelValues {
+  const presetModelValues = PRESET_MODELS.filter(
+    (modelValues: ModelValues): boolean =>
+      modelValues.checkpointUrl === checkpointUrl
+  );
+  return presetModelValues.length > 0 ? presetModelValues[0] : {};
+}
+
+async function synthesize(
```

Comment:
Since this function only is used in the context of a model, wouldn't it make sense to just have it as a method of the Model class, rather than a separate function?

Reply:
this is a method of the model class, or maybe i am missing your point. the only thing i did here is to separate it out into it's own file to keep the model class a little bit more lightweight. i can move it all into the same file.

Reply:
yah, that's it. It just makes sense to have all the code together, because these are all the things specific to the model. (i.e. `model` is always `self`)